// safeness-sb-new/src/client/websocket/WebSocketManager.js
'use strict';

const EventEmitter = require('node:events');
const { setImmediate } = require('node:timers');
const { setTimeout: sleep } = require('node:timers/promises');
const { Collection } = require('@discordjs/collection');
const { RPCErrorCodes } = require('discord-api-types/v9');
const WebSocketShard = require('./WebSocketShard');
const PacketHandlers = require('./handlers');
const { Error } = require('../../errors');
const { Events, ShardEvents, Status, WSCodes, WSEvents } = require('../../util/Constants');

const BeforeReadyWhitelist = [
  WSEvents.READY,
  WSEvents.RESUMED,
  WSEvents.GUILD_CREATE,
  WSEvents.GUILD_DELETE,
  WSEvents.GUILD_MEMBERS_CHUNK,
  WSEvents.GUILD_MEMBER_ADD,
  WSEvents.GUILD_MEMBER_REMOVE,
];

const UNRECOVERABLE_CLOSE_CODES = Object.keys(WSCodes).slice(2).map(Number);
const UNRESUMABLE_CLOSE_CODES = [
  RPCErrorCodes.UnknownError,
  RPCErrorCodes.InvalidPermissions,
  RPCErrorCodes.InvalidClientId,
];

class WebSocketManager extends EventEmitter {
  constructor(client) {
    super();

    Object.defineProperty(this, 'client', { value: client });

    this.gateway = null;
    this.totalShards = this.client.options.shards.length;
    this.shards = new Collection();

    Object.defineProperty(this, 'shardQueue', { value: new Set(), writable: true });
    Object.defineProperty(this, 'packetQueue', { value: [] });
    this.status = Status.IDLE;
    this.destroyed = false;
    this.reconnecting = false;

    Object.defineProperty(this, 'connectionConfig', {
      value: {
        shardDelay: { min: 5000, max: 7500 },
        reconnectDelay: { min: 1000, max: 3000 },
        maxConcurrentConnections: 1,
        sessionPreservation: true,
        adaptiveRateLimiting: true,
      },
      writable: true,
    });

    Object.defineProperty(this, 'activeConnections', { value: new Set(), writable: true });
    Object.defineProperty(this, 'connectionHistory', { value: [], writable: true });
    Object.defineProperty(this, 'lastActivity', { value: 0, writable: true });
  }

  getRandomDelay(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  get ping() {
    const sum = this.shards.reduce((a, b) => a + b.ping, 0);
    return sum / this.shards.size;
  }

  debug(message, shard) {
    this.client.emit(Events.DEBUG, `[WS => ${shard ? `Shard ${shard.id}` : 'Manager'}] ${message}`);
  }

  logConnectionAttempt(shardId, success) {
    this.connectionHistory.push({
      timestamp: Date.now(),
      shardId,
      success,
    });

    if (this.connectionHistory.length > 100) {
      this.connectionHistory.shift();
    }

    this.lastActivity = Date.now();
  }

  calculateBackoff(shardId) {
    const now = Date.now();
    const recentFailures = this.connectionHistory.filter(
      entry => entry.shardId === shardId && !entry.success && now - entry.timestamp < 60000
    ).length;

    if (recentFailures === 0) return 0;

    const backoffTime = Math.min(1000 * Math.pow(2, recentFailures), 30000);
    const jitter = this.getRandomDelay(-500, 500);
    
    return backoffTime + jitter;
  }

  async connect() {
    await sleep(this.getRandomDelay(500, 1500));

    let gatewayURL = 'wss://gateway.discord.gg';
    
    await sleep(this.getRandomDelay(200, 800));
    
    try {
      await this.client.api.gateway
        .get({ auth: false })
        .then(r => (gatewayURL = r.url))
        .catch(() => {});
    } catch (error) {
      this.debug('Échec récupération gateway, nouvelle tentative...');
      await sleep(2000);
      return this.connect();
    }

    const total = Infinity;
    const remaining = Infinity;
    const recommendedShards = 1;

    this.gateway = `${gatewayURL}/`;

    let { shards } = this.client.options;

    if (shards === 'auto') {
      this.totalShards = this.client.options.shardCount = recommendedShards;
      shards = this.client.options.shards = Array.from({ length: recommendedShards }, (_, i) => i);
    }

    this.totalShards = shards.length;
    
    const shardInstances = [];
    for (let i = 0; i < shards.length; i++) {
      await sleep(this.getRandomDelay(100, 300));
      shardInstances.push(new WebSocketShard(this, shards[i]));
    }
    
    this.shardQueue = new Set(shardInstances);

    return this.createShards();
  }

  async createShards() {
    if (!this.shardQueue.size) {
      this.activeConnections.clear();
      return false;
    }

    if (this.activeConnections.size >= this.connectionConfig.maxConcurrentConnections) {
      this.debug(`Limite connexions atteinte, attente...`);
      await sleep(1000);
      return this.createShards();
    }

    const [shard] = this.shardQueue;
    this.shardQueue.delete(shard);
    this.activeConnections.add(shard);

    const backoffDelay = this.calculateBackoff(shard.id);
    if (backoffDelay > 0) {
      this.debug(`Backoff ${backoffDelay}ms shard ${shard.id}`);
      await sleep(backoffDelay);
    }

    if (!shard.eventsAttached) {
      shard.on(ShardEvents.ALL_READY, unavailableGuilds => {
        this.client.emit(Events.SHARD_READY, shard.id, unavailableGuilds);
        this.logConnectionAttempt(shard.id, true);
        this.activeConnections.delete(shard);

        if (!this.shardQueue.size) this.reconnecting = false;
        this.checkShardsReady();
      });

      shard.on(ShardEvents.CLOSE, event => {
        this.activeConnections.delete(shard);
        this.logConnectionAttempt(shard.id, false);

        if (event.code === 1_000 ? this.destroyed : UNRECOVERABLE_CLOSE_CODES.includes(event.code)) {
          this.client.emit(Events.SHARD_DISCONNECT, event, shard.id);
          return;
        }

        if (UNRESUMABLE_CLOSE_CODES.includes(event.code)) {
          shard.sessionId = null;
        }

        this.client.emit(Events.SHARD_RECONNECTING, shard.id);

        const reconnectDelay = this.getRandomDelay(
          this.connectionConfig.reconnectDelay.min,
          this.connectionConfig.reconnectDelay.max
        );
        
        setTimeout(() => {
          this.shardQueue.add(shard);
          
          if (shard.sessionId && this.connectionConfig.sessionPreservation) {
            this.debug('Session présente, reconnexion...', shard);
          }
          
          this.reconnect();
        }, reconnectDelay);
      });

      shard.on(ShardEvents.INVALID_SESSION, () => {
        this.client.emit(Events.SHARD_RECONNECTING, shard.id);
      });

      shard.on(ShardEvents.DESTROYED, () => {
        this.debug('Shard détruit, reconnexion...', shard);
        this.client.emit(Events.SHARD_RECONNECTING, shard.id);

        const reconnectDelay = this.getRandomDelay(1000, 3000);
        
        setTimeout(() => {
          this.shardQueue.add(shard);
          this.reconnect();
        }, reconnectDelay);
      });

      shard.eventsAttached = true;
    }

    this.shards.set(shard.id, shard);

    try {
      const connectDelay = this.getRandomDelay(200, 800);
      await sleep(connectDelay);
      
      await shard.connect();
    } catch (error) {
      this.activeConnections.delete(shard);
      this.logConnectionAttempt(shard.id, false);
      
      if (error?.code && UNRECOVERABLE_CLOSE_CODES.includes(error.code)) {
        throw new Error(WSCodes[error.code]);
      } else if (!error || error.code) {
        this.debug('Échec connexion gateway, remise file d\'attente...', shard);
        
        const requeueDelay = this.getRandomDelay(1000, 3000);
        await sleep(requeueDelay);
        
        this.shardQueue.add(shard);
      } else {
        throw error;
      }
    }
    
    if (this.shardQueue.size) {
      const delay = this.getRandomDelay(
        this.connectionConfig.shardDelay.min,
        this.connectionConfig.shardDelay.max
      );
      
      this.debug(`File d\'attente shards: ${this.shardQueue.size}; attente ${delay}ms`);
      await sleep(delay);
      return this.createShards();
    }

    return true;
  }

  async reconnect() {
    if (this.reconnecting || this.status !== Status.READY) return false;
    
    const startDelay = this.getRandomDelay(500, 1500);
    await sleep(startDelay);
    
    this.reconnecting = true;
    
    try {
      await this.createShards();
    } catch (error) {
      this.debug(`Impossible reconnexion: ${error}`);
      
      if (error.httpStatus !== 401) {
        const retryDelay = this.getRandomDelay(3000, 8000);
        this.debug(`Erreur réseau, nouvelle tentative ${retryDelay}ms`);
        await sleep(retryDelay);
        this.reconnecting = false;
        return this.reconnect();
      }
      
      if (this.client.listenerCount(Events.INVALIDATED)) {
        this.client.emit(Events.INVALIDATED);
        this.destroy();
      } else {
        this.client.destroy();
      }
    } finally {
      this.reconnecting = false;
    }
    
    return true;
  }

  broadcast(packet) {
    let delay = 0;
    
    for (const shard of this.shards.values()) {
      setTimeout(() => {
        shard.send(packet);
      }, delay);
      
      delay += this.getRandomDelay(10, 50);
    }
    
    this.lastActivity = Date.now();
  }

  destroy() {
    if (this.destroyed) return;
    
    this.destroyed = true;
    this.shardQueue.clear();
    this.activeConnections.clear();
    
    let delay = 0;
    const shardsArray = Array.from(this.shards.values());
    
    for (const shard of shardsArray) {
      setTimeout(() => {
        shard.destroy({ closeCode: 1_000, reset: true, emit: false, log: false });
      }, delay);
      
      delay += this.getRandomDelay(100, 300);
    }
  }

  handlePacket(packet, shard) {
    this.lastActivity = Date.now();
    
    if (packet && this.status !== Status.READY) {
      if (!BeforeReadyWhitelist.includes(packet.t)) {
        const queueDelay = this.getRandomDelay(50, 200);
        
        setTimeout(() => {
          this.packetQueue.push({ packet, shard });
        }, queueDelay);
        
        return false;
      }
    }

    if (this.packetQueue.length) {
      const processWithDelay = (index) => {
        if (index >= this.packetQueue.length) return;
        
        const item = this.packetQueue[index];
        const delay = this.getRandomDelay(10, 50);
        
        setTimeout(() => {
          this.handlePacket(item.packet, item.shard);
          processWithDelay(index + 1);
        }, delay);
      };
      
      processWithDelay(0);
      this.packetQueue.length = 0;
    }

    if (packet && PacketHandlers[packet.t]) {
      const handlerDelay = this.getRandomDelay(5, 20);
      
      setTimeout(() => {
        PacketHandlers[packet.t](this.client, packet, shard);
      }, handlerDelay);
    } else if (packet) {
      const emitDelay = this.getRandomDelay(5, 15);
      
      setTimeout(() => {
        this.client.emit(Events.UNHANDLED_PACKET, packet, shard);
      }, emitDelay);
    }

    return true;
  }

  checkShardsReady() {
    if (this.status === Status.READY) return;
    
    const allShardsReady = this.shards.size === this.totalShards && 
                          this.shards.every(s => s.status === Status.READY);
    
    if (!allShardsReady) {
      const checkDelay = this.getRandomDelay(100, 500);
      
      setTimeout(() => {
        this.checkShardsReady();
      }, checkDelay);
      
      return;
    }

    const readyDelay = this.getRandomDelay(200, 800);
    
    setTimeout(() => {
      this.triggerClientReady();
    }, readyDelay);
  }

  triggerClientReady() {
    this.status = Status.READY;
    this.client.readyAt = new Date();
    this.client.emit(Events.CLIENT_READY, this.client);
    this.handlePacket();
  }
}

module.exports = WebSocketManager;
