// safeness-sb-new/src/client/websocket/WebSocketShard.js
'use strict';

const EventEmitter = require('node:events');
const { setTimeout, setInterval, clearTimeout } = require('node:timers');
const WebSocket = require('../../WebSocket');
const { Status, Events, ShardEvents, Opcodes, WSEvents, WSCodes } = require('../../util/Constants');
const Intents = require('../../util/Intents');
const Util = require('../../util/Util');

const STATUS_KEYS = Object.keys(Status);
const CONNECTION_STATE = Object.keys(WebSocket.WebSocket);

let zlib;

try {
  zlib = require('zlib-sync');
} catch {} // eslint-disable-line no-empty

class WebSocketShard extends EventEmitter {
  constructor(manager, id) {
    super();

    this.manager = manager;
    this.id = id;
    this.resumeURL = null;
    this.status = Status.IDLE;
    this.sequence = -1;
    this.closeSequence = 0;
    this.sessionId = null;
    this.ping = -1;
    this.lastPingTimestamp = -1;
    this.lastHeartbeatAcked = true;
    this.closeEmitted = false;

    Object.defineProperty(this, 'ratelimit', {
      value: {
        queue: [],
        total: 115,
        remaining: 115,
        time: 60e3,
        timer: null,
        lastReset: Date.now(),
      },
    });

    Object.defineProperty(this, 'connection', { value: null, writable: true });
    Object.defineProperty(this, 'inflate', { value: null, writable: true });
    Object.defineProperty(this, 'helloTimeout', { value: null, writable: true });
    Object.defineProperty(this, 'wsCloseTimeout', { value: null, writable: true });
    Object.defineProperty(this, 'eventsAttached', { value: false, writable: true });
    Object.defineProperty(this, 'expectedGuilds', { value: null, writable: true });
    Object.defineProperty(this, 'readyTimeout', { value: null, writable: true });
    Object.defineProperty(this, 'connectedAt', { value: 0, writable: true });

    Object.defineProperty(this, 'humanDelays', {
      value: {
        reconnect: { min: 1000, max: 5000 },
        identify: { min: 5000, max: 10000 },
        heartbeatJitter: { min: -100, max: 100 },
      },
      writable: true,
    });

    Object.defineProperty(this, 'connectionAttempts', { value: 0, writable: true });
    Object.defineProperty(this, 'lastActivity', { value: 0, writable: true });
  }

  getRandomDelay(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  debug(message) {
    const ignoredDebugs = [
      'Heartbeat accusé réception',
      '[HeartbeatTimer] Envoi d\'un heartbeat',
      'Effacement du timeout HELLO',
      'Définition d\'un intervalle de heartbeat'
    ];
    
    if (!ignoredDebugs.some(ignored => message.includes(ignored))) {
      this.manager.debug(message, this);
    }
  }

  connect() {
    const { client } = this.manager;

    if (this.connection?.readyState === WebSocket.OPEN && this.status === Status.READY) {
      return Promise.resolve();
    }

    const gateway = this.resumeURL ?? this.manager.gateway;

    return new Promise((resolve, reject) => {
      const cleanup = () => {
        this.removeListener(ShardEvents.CLOSE, onClose);
        this.removeListener(ShardEvents.READY, onReady);
        this.removeListener(ShardEvents.RESUMED, onResumed);
        this.removeListener(ShardEvents.INVALID_SESSION, onInvalidOrDestroyed);
        this.removeListener(ShardEvents.DESTROYED, onInvalidOrDestroyed);
      };

      const onReady = () => {
        cleanup();
        resolve();
      };

      const onResumed = () => {
        cleanup();
        resolve();
      };

      const onClose = event => {
        cleanup();
        reject(event);
      };

      const onInvalidOrDestroyed = () => {
        cleanup();
        reject();
      };

      this.once(ShardEvents.READY, onReady);
      this.once(ShardEvents.RESUMED, onResumed);
      this.once(ShardEvents.CLOSE, onClose);
      this.once(ShardEvents.INVALID_SESSION, onInvalidOrDestroyed);
      this.once(ShardEvents.DESTROYED, onInvalidOrDestroyed);

      if (this.connection?.readyState === WebSocket.OPEN) {
        this.debug('Connexion existante détectée.');
        this.identify();
        return;
      }

      if (this.connection) {
        this.debug(`Nettoyage connexion précédente. État: ${CONNECTION_STATE[this.connection.readyState]}`);
        this.destroy({ emit: false });
      }

      const wsQuery = { v: client.options.ws.version };

      if (zlib) {
        this.inflate = new zlib.Inflate({
          chunkSize: 65535,
          flush: zlib.Z_SYNC_FLUSH,
          to: WebSocket.encoding === 'json' ? 'string' : '',
        });
        wsQuery.compress = 'zlib-stream';
      }

      this.debug(`Initialisation connexion: ${gateway}`);

      this.status = this.status === Status.DISCONNECTED ? Status.RECONNECTING : Status.CONNECTING;
      this.setHelloTimeout();
      this.setWsCloseTimeout(-1);
      this.connectedAt = Date.now();

      const reconnectDelay = this.getRandomDelay(
        this.humanDelays.reconnect.min,
        this.humanDelays.reconnect.max
      );
      
      this.debug(`Délai connexion: ${reconnectDelay}ms`);
      
      setTimeout(() => {
        const ws = (this.connection = WebSocket.create(gateway, wsQuery, {
          handshakeTimeout: 30_000,
          agent: Util.verifyProxyAgent(client.options.ws.agent) ? client.options.ws.agent : undefined,
        }));
        ws.onopen = this.onOpen.bind(this);
        ws.onmessage = this.onMessage.bind(this);
        ws.onerror = this.onError.bind(this);
        ws.onclose = this.onClose.bind(this);
        this.connectionAttempts++;
      }, reconnectDelay);
    });
  }

  onOpen() {
    const connectTime = Date.now() - this.connectedAt;
    if (connectTime > 5000 || connectTime < 100) {
      this.debug(`Connexion établie: ${connectTime}ms`);
    }
    this.status = Status.NEARLY;
    this.connectionAttempts = 0;
  }

  onMessage({ data }) {
    let raw;
    if (data instanceof ArrayBuffer) data = new Uint8Array(data);
    if (zlib) {
      const l = data.length;
      const flush =
        l >= 4 && data[l - 4] === 0x00 && data[l - 3] === 0x00 && data[l - 2] === 0xff && data[l - 1] === 0xff;

      this.inflate.push(data, flush && zlib.Z_SYNC_FLUSH);
      if (!flush) return;
      raw = this.inflate.result;
    } else {
      raw = data;
    }
    let packet;
    try {
      packet = WebSocket.unpack(raw);
    } catch (err) {
      this.manager.client.emit(Events.SHARD_ERROR, err, this.id);
      return;
    } 
    
    this.onPacket(packet);
    
    if (packet.op === Opcodes.DISPATCH && packet.t) {
      const essentialEvents = [
        'READY',
        'RESUMED', 
        'GUILD_MEMBERS_CHUNK',
        'MESSAGE_CREATE',
        'MESSAGE_DELETE', 
        'GUILD_CREATE',
        'PRESENCE_UPDATE',
        'CHANNEL_CREATE',
        'CHANNEL_DELETE',
        'THREAD_CREATE',
        'GUILD_ROLE_CREATE',
        'GUILD_ROLE_DELETE',
        'GUILD_ROLE_UPDATE',
        'CHANNEL_RECIPIENT_ADD',
        'CHANNEL_RECIPIENT_REMOVE',
        'VOICE_STATE_UPDATE',
        'VOICE_SERVER_UPDATE',
        'VOICE_CHANNEL_STATUS_UPDATE',
        'INTERACTION_CREATE',
        'VOICE_STATE_UPDATE',
        'USER_UPDATE'
      ];
      
      if (essentialEvents.includes(packet.t)) {
        this.manager.emit(packet.t, packet.d, this.id);
      }
    }
  }

  onError(event) {
    const error = event?.error ?? event;
    if (!error) return;
    this.manager.client.emit(Events.SHARD_ERROR, error, this.id);
  }

  onClose(event) {
    this.closeEmitted = true;
    if (this.sequence !== -1) this.closeSequence = this.sequence;
    this.sequence = -1;
    this.setHeartbeatTimer(-1);
    this.setHelloTimeout(-1);
    this.setWsCloseTimeout(-1);
    if (this.connection) {
      this._cleanupConnection();
      this.destroy({ reset: !this.sessionId, emit: false, log: false });
    }
    this.status = Status.DISCONNECTED;
    this.emitClose(event);
  }

  emitClose(
    event = {
      code: 1011,
      reason: WSCodes[1011],
      wasClean: false,
    },
  ) {
    this.debug(`Fermeture: ${event.code}`);
    this.emit(ShardEvents.CLOSE, event);
  }

  onPacket(packet) {
    if (!packet) {
      this.debug(`Paquet invalide: '${packet}'.`);
      return;
    }

    switch (packet.t) {
      case WSEvents.READY:
        this.emit(ShardEvents.READY);
        this.resumeURL = packet.d.resume_gateway_url;
        this.sessionId = packet.d.session_id;
        this.expectedGuilds = new Set(packet.d.guilds.filter(d => d?.unavailable == true).map(d => d.id));
        this.status = Status.WAITING_FOR_GUILDS;
        this.debug(`Session ${this.sessionId} établie.`);
        this.lastHeartbeatAcked = true;
        this.sendHeartbeat('ReadyHeartbeat');
        break;
      case WSEvents.RESUMED: {
        this.emit(ShardEvents.RESUMED);
        this.status = Status.READY;
        const replayed = packet.s - this.closeSequence;
        this.debug(`Session ${this.sessionId} reprise.`);
        this.lastHeartbeatAcked = true;
        this.sendHeartbeat('ResumeHeartbeat');
        break;
      }
    }

    if (packet.s > this.sequence) this.sequence = packet.s;

    switch (packet.op) {
      case Opcodes.HELLO:
        this.setHelloTimeout(-1);
        this.setHeartbeatTimer(packet.d.heartbeat_interval);
        const identifyDelay = this.getRandomDelay(
          this.humanDelays.identify.min,
          this.humanDelays.identify.max
        );
        this.debug(`Délai identification: ${identifyDelay}ms`);
        setTimeout(() => this.identify(), identifyDelay);
        break;
      case Opcodes.RECONNECT:
        this.debug('Reconnexion demandée.');
        this.destroy({ closeCode: 4_000 });
        break;
      case Opcodes.INVALID_SESSION:
        this.debug(`Session invalide: ${packet.d}.`);
        if (packet.d) {
          this.identifyResume();
          return;
        }
        this.sequence = -1;
        this.sessionId = null;
        this.status = Status.RECONNECTING;
        this.emit(ShardEvents.INVALID_SESSION);
        break;
      case Opcodes.HEARTBEAT_ACK:
        this.ackHeartbeat();
        break;
      case Opcodes.HEARTBEAT:
        this.sendHeartbeat('HeartbeatRequest', true);
        break;
      default:
        this.manager.handlePacket(packet, this);
        if (this.status === Status.WAITING_FOR_GUILDS && packet.t === WSEvents.GUILD_CREATE) {
          this.expectedGuilds.delete(packet.d.id);
          this.checkReady();
        }
    }
  }

  checkReady() {
    if (this.readyTimeout) {
      clearTimeout(this.readyTimeout);
      this.readyTimeout = null;
    }
    if (!this.expectedGuilds.size) {
      this.debug('Tous serveurs reçus.');
      this.status = Status.READY;
      this.emit(ShardEvents.ALL_READY);
      return;
    }
    const hasGuildsIntent = new Intents(this.manager.client.options.intents).has(Intents.FLAGS.GUILDS);
    const { waitGuildTimeout } = this.manager.client.options;

    this.readyTimeout = setTimeout(
      () => {
        this.debug(`Serveurs manquants: ${this.expectedGuilds.size}`);
        this.readyTimeout = null;
        this.status = Status.READY;
        this.emit(ShardEvents.ALL_READY, this.expectedGuilds);
      },
      hasGuildsIntent ? waitGuildTimeout : 0,
    ).unref();
  }

  setHelloTimeout(time) {
    if (time === -1) {
      if (this.helloTimeout) {
        clearTimeout(this.helloTimeout);
        this.helloTimeout = null;
      }
      return;
    }
    this.helloTimeout = setTimeout(() => {
      this.debug('Timeout connexion.');
      this.destroy({ reset: true, closeCode: 4009 });
    }, 20_000).unref();
  }

  setWsCloseTimeout(time) {
    if (this.wsCloseTimeout) {
      clearTimeout(this.wsCloseTimeout);
    }
    if (time === -1) {
      this.wsCloseTimeout = null;
      return;
    }
    this.wsCloseTimeout = setTimeout(() => {
      this.setWsCloseTimeout(-1);
      if (this.closeEmitted) {
        this.closeEmitted = false;
        return;
      }
      if (this.connection) this._cleanupConnection();
      this.emitClose({
        code: 4009,
        reason: 'Session expirée.',
        wasClean: false,
      });
    }, time);
  }

  setHeartbeatTimer(time) {
    if (time === -1) {
      if (this.heartbeatInterval) {
        clearInterval(this.heartbeatInterval);
        this.heartbeatInterval = null;
      }
      return;
    }
    
    const jitter = this.getRandomDelay(
      this.humanDelays.heartbeatJitter.min,
      this.humanDelays.heartbeatJitter.max
    );
    const adjustedTime = time + jitter;
    
    if (this.heartbeatInterval) clearInterval(this.heartbeatInterval);
    this.heartbeatInterval = setInterval(() => this.sendHeartbeat(), adjustedTime).unref();
  }

  sendHeartbeat(
    tag = 'HeartbeatTimer',
    ignoreHeartbeatAck = [Status.WAITING_FOR_GUILDS, Status.IDENTIFYING, Status.RESUMING].includes(this.status),
  ) {
    if (ignoreHeartbeatAck && !this.lastHeartbeatAcked) {
      this.debug(`[${tag}] Envoi heartbeat forcé.`);
    } else if (!this.lastHeartbeatAcked) {
      this.debug('Heartbeat non reçu. Reconnexion.');
      this.destroy({ reset: true, closeCode: 4009 });
      return;
    }

    const now = Date.now();
    const timeSinceLastActivity = now - this.lastActivity;
    
    if (timeSinceLastActivity < 100) {
      setTimeout(() => {
        this.sendHeartbeat(tag, ignoreHeartbeatAck);
      }, 100 - timeSinceLastActivity);
      return;
    }

    this.lastHeartbeatAcked = false;
    this.lastPingTimestamp = Date.now();
    this.lastActivity = Date.now();
    this.send({ op: Opcodes.HEARTBEAT, d: this.sequence }, true);
  }

  ackHeartbeat() {
    this.lastHeartbeatAcked = true;
    const latency = Date.now() - this.lastPingTimestamp;
    this.debug(`Latence: ${latency}ms.`);
    this.ping = latency;
  }

  identify() {
    return this.sessionId ? this.identifyResume() : this.identifyNew();
  }

  _getIdentifyProperties() {
    const { client } = this.manager;
    const wsProps = client.options.ws.properties || {};
    
    return {
      os: wsProps.os || 'windows',
      browser: wsProps.browser || 'Discord Client',
      device: wsProps.browser || 'Discord Client',
      system_locale: wsProps.system_locale || 'fr-FR',
      browser_user_agent: wsProps.browser_user_agent || '',
      browser_version: wsProps.browser_version || '',
      os_version: wsProps.os_version || '10.0.19045',
      referrer: '',
      referring_domain: '',
      referrer_current: '',
      referring_domain_current: '',
      release_channel: wsProps.release_channel || 'stable',
      client_build_number: wsProps.client_build_number || 297020,
      client_event_source: wsProps.client_event_source || null,
      design_id: wsProps.design_id || 0
    };
  }

  identifyNew() {
    const { client } = this.manager;
    if (!client.token) {
      this.debug('Token manquant.');
      return;
    }

    if (this.connectionAttempts > 3) {
      const backoffTime = Math.min(1000 * Math.pow(2, this.connectionAttempts - 3), 30000);
      this.debug(`Backoff: ${backoffTime}ms`);
      setTimeout(() => {
        this.identifyNew();
      }, backoffTime);
      return;
    }

    this.status = Status.IDENTIFYING;

    const wsProperties = this._getIdentifyProperties();
    
    const d = {
      token: client.token,
      properties: wsProperties,
      compress: client.options.ws.compress || false,
      large_threshold: client.options.ws.large_threshold || 50,
      shard: client.options.shard || [0, 1],
      presence: client.options.presence || {},
      intents: client.options.intents || 0
    };

    delete d.version;
    delete d.agent;

    this.debug(`Identification shard ${this.id}`);
    
    const sendDelay = this.getRandomDelay(100, 500);
    setTimeout(() => {
      this.send({ op: Opcodes.IDENTIFY, d }, true);
    }, sendDelay);
  }

  identifyResume() {
    if (!this.sessionId) {
      this.debug('Reprise sans session.');
      this.identifyNew();
      return;
    }

    this.status = Status.RESUMING;
    this.debug(`Reprise session ${this.sessionId}`);

    const d = {
      token: this.manager.client.token,
      session_id: this.sessionId,
      seq: this.closeSequence,
    };

    const sendDelay = this.getRandomDelay(100, 500);
    setTimeout(() => {
      this.send({ op: Opcodes.RESUME, d }, true);
    }, sendDelay);
  }

  send(data, important = false) {
    const now = Date.now();
    const timeSinceLastActivity = now - this.lastActivity;
    
    if (timeSinceLastActivity < 50 && !important) {
      setTimeout(() => {
        this.send(data, important);
      }, 50 - timeSinceLastActivity);
      return;
    }

    this.ratelimit.queue[important ? 'unshift' : 'push'](data);
    this.processQueue();
    this.lastActivity = Date.now();
  }

  _send(data) {
    if (this.connection?.readyState !== WebSocket.OPEN) {
      this.debug('Envoi impossible: connexion fermée.');
      this.destroy({ closeCode: 4_000 });
      return;
    }

    this.connection.send(WebSocket.pack(data), err => {
      if (err) this.manager.client.emit(Events.SHARD_ERROR, err, this.id);
    });
  }

  processQueue() {
    if (this.ratelimit.remaining === 0) return;
    if (this.ratelimit.queue.length === 0) return;
    
    const now = Date.now();
    if (now - this.ratelimit.lastReset >= this.ratelimit.time) {
      this.ratelimit.remaining = this.ratelimit.total;
      this.ratelimit.lastReset = now;
    }
    
    if (this.ratelimit.remaining === this.ratelimit.total) {
      this.ratelimit.timer = setTimeout(() => {
        this.ratelimit.remaining = this.ratelimit.total;
        this.ratelimit.lastReset = Date.now();
        this.processQueue();
      }, this.ratelimit.time).unref();
    }
    
    const maxPacketsPerCycle = 5;
    let packetsSent = 0;
    
    while (this.ratelimit.remaining > 0 && packetsSent < maxPacketsPerCycle) {
      const item = this.ratelimit.queue.shift();
      if (!item) return;
      
      if (packetsSent > 0) {
        setTimeout(() => {
          this._send(item);
        }, packetsSent * 10);
      } else {
        this._send(item);
      }
      
      this.ratelimit.remaining--;
      packetsSent++;
    }
    
    if (this.ratelimit.queue.length > 0) {
      setTimeout(() => {
        this.processQueue();
      }, 50);
    }
  }

  destroy({ closeCode = 1_000, reset = false, emit = true, log = true } = {}) {
    if (log) {
      this.debug(`Destruction: ${closeCode}`);
    }

    this.setHeartbeatTimer(-1);
    this.setHelloTimeout(-1);

    if (this.connection) {
      if (this.connection.readyState === WebSocket.OPEN) {
        this.connection.close(closeCode);
      } else {
        try {
          this.connection.close(closeCode);
        } catch (err) {
          this.connection.terminate();
        }
        if (emit) this._emitDestroyed();
      }
    } else if (emit) {
      this._emitDestroyed();
    }

    this.setWsCloseTimeout(this.manager.client.options.closeTimeout);
    this.connection = null;
    this.status = Status.DISCONNECTED;

    if (this.sequence !== -1) this.closeSequence = this.sequence;

    if (reset) {
      this.resumeURL = null;
      this.sequence = -1;
      this.sessionId = null;
    }

    this.ratelimit.remaining = this.ratelimit.total;
    this.ratelimit.queue.length = 0;
    if (this.ratelimit.timer) {
      clearTimeout(this.ratelimit.timer);
      this.ratelimit.timer = null;
    }
  }

  _cleanupConnection() {
    if (this.connection) {
      this.connection.onopen = this.connection.onclose = this.connection.onmessage = null;
      this.connection.onerror = () => null;
    }
  }

  _emitDestroyed() {
    this.emit(ShardEvents.DESTROYED);
  }
}

module.exports = WebSocketShard;
